README.md
# RobotlegsJS PixiJS Extension

[![GitHub license](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobotlegsJS/RobotlegsJS-Pixi/blob/master/LICENSE)
[![Gitter chat](https://badges.gitter.im/RobotlegsJS/RobotlegsJS.svg)](https://gitter.im/RobotlegsJS/RobotlegsJS)
[![Build Status](https://travis-ci.com/RobotlegsJS/RobotlegsJS-Pixi.svg?branch=master)](https://travis-ci.com/RobotlegsJS/RobotlegsJS-Pixi)
[![codebeat badge](https://codebeat.co/badges/f9e9f52e-ddef-4f51-82b7-9e6a695474ab)](https://codebeat.co/projects/github-com-robotlegsjs-robotlegsjs-pixi-master)
[![Test Coverage](https://codeclimate.com/github/RobotlegsJS/RobotlegsJS-Pixi/badges/coverage.svg)](https://codeclimate.com/github/RobotlegsJS/RobotlegsJS-Pixi/coverage)
[![npm version](https://badge.fury.io/js/%40robotlegsjs%2Fpixi.svg)](https://badge.fury.io/js/%40robotlegsjs%2Fpixi)
[![Greenkeeper badge](https://badges.greenkeeper.io/RobotlegsJS/RobotlegsJS-Pixi.svg)](https://greenkeeper.io/)
[![styled with prettier](https://img.shields.io/badge/styled_with-prettier-ff69b4.svg)](https://github.com/prettier/prettier)

Integrate [RobotlegsJS](https://github.com/RobotlegsJS/RobotlegsJS)
framework with [PixiJS](https://github.com/pixijs/pixi.js).

## Installation

### Get `@robotlegsjs/pixi`

You can get the latest release and the type definitions using [NPM](https://www.npmjs.com/):

```bash
npm install @robotlegsjs/pixi --save
```

Or using [Yarn](https://yarnpkg.com/en/):

```bash
yarn add @robotlegsjs/pixi
```

From version `0.2.0` of this package, the [PixiJS](https://github.com/pixijs/pixi.js) dependencies were moved to **peerDependencies**,
allowing the final user to choose the desired version of the `pixi.js` library on each project.

#### For `pixi.js` version 5

The `@robotlegsjs/pixi` package version `^2.0.0` is compatible with versions `>=5.0.0 <6` version range of `pixi.js` library.

Since version `5.0.0` of `pixi.js`, the `eventemitter3` library was removed and is not necessary anymore.

You can setup your project for version 5 of `pixi.js` using:

```bash
npm install @robotlegsjs/pixi@^2.0.0 pixi.js@^5.0.0 --save
```

or

```bash
yarn add @robotlegsjs/pixi@^2.0.0 pixi.js@^5.0.0
```

#### For `pixi.js` version 4

For `pixi.js` `>=4.2.1 <5`, please use `@robotlegsjs/pixi` version `^1`.

```bash
npm install @robotlegsjs/pixi@^1.0.1 --save
```

or

```bash
yarn add @robotlegsjs/pixi@^1.0.1
```

Since each version of `pixi.js` v4 library defines which version of `eventemitter3` library is being used, remember to also install the proper version of `eventemitter3` in your project.

As example, when you would like to use the version `4.2.1` of `pixi.js` library, you can run:

```bash
npm install pixi.js@4.2.1 eventemitter3@^2.0.0 reflect-metadata --save
```

or

```bash
yarn add pixi.js@4.2.1 eventemitter3@^2.0.0 reflect-metadata
```

### Setup of your project

Then follow the [installation instructions](https://github.com/RobotlegsJS/RobotlegsJS/blob/master/README.md#installation) of **RobotlegsJS** library to complete the setup of your project.

**Dependencies**

+ [RobotlegsJS](https://github.com/RobotlegsJS/RobotlegsJS)
+ [tslib](https://github.com/Microsoft/tslib)

**Peer Dependencies**

+ [PixiJS](https://github.com/pixijs/pixi.js)
+ [eventemitter3](https://github.com/primus/eventemitter3)
+ [reflect-metadata](https://github.com/rbuckton/reflect-metadata)

## Usage

```typescript
import "reflect-metadata";

import * as PIXI from 'pixi.js';

import { Context, MVCSBundle } from "@robotlegsjs/core";
import { ContextView, PixiBundle } from "@robotlegsjs/pixi";

import { MyConfig } from "./config/MyConfig";
import { RobotlegsView } from "./view/RobotlegsView";

export class Game {

    private canvas: HTMLCanvasElement;
    private stage: PIXI.Container;
    private renderer: PIXI.Renderer;
    private context: Context;

    constructor () {
        this.canvas = <HTMLCanvasElement>(document.getElementById("canvas"));
        this.renderer = PIXI.autoDetectRenderer({
            width: 960,
            height: 400,
            view: this.canvas,
            backgroundColor: 0xffffff
        });
        this.stage = new PIXI.Container();

        this.context = new Context();
        this.context.install(MVCSBundle, PixiBundle).
            configure(new ContextView(this.stage)).
            configure(MyConfig).
            initialize();

        this.stage.addChild(new RobotlegsView());

        document.body.appendChild(this.renderer.view);

        this.render();
    }

    public render = () => {
        this.renderer.render(this.stage);
        window.requestAnimationFrame(this.render);
    }

}

let game: Game = new Game();

```

[See full example here](example/index.ts)

## Running the example

Run the following commands to run the example:

```bash
npm install
npm start
```

or:

```bash
yarn install
yarn start
```

## Checking the example

You can check our example project [here](http://robotlegsjs.io/RobotlegsJS-Pixi).

## RobotlegsJS PixiJS for enterprise

Available as part of the Tidelift Subscription

The maintainers of [@robotlegsjs/pixi](https://github.com/RobotlegsJS/RobotlegsJS-Pixi) and thousands of other packages are working with Tidelift to deliver commercial support and maintenance for the open source dependencies you use to build your applications. Save time, reduce risk, and improve code health, while paying the maintainers of the exact dependencies you use. [Learn more.](https://tidelift.com/subscription/pkg/npm-robotlegsjs-pixi?utm_source=npm-robotlegsjs-pixi&utm_medium=referral&utm_campaign=enterprise&utm_term=repo)

## License

[MIT](LICENSE)


CHANGELOG.md
# RobotlegsJS Pixi Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Suggestions or improvements for further versions

- [x] Add instructions of how to install the **@robotlegsjs/pixi** package into **README.md**.

- [ ] Use [**Function Types**](https://www.typescriptlang.org/docs/handbook/functions.html) for handlers and callbacks instead of generic **Function** type.

- [ ] Evaluate if **IMediator** interface should be mandatory.

- [x] Update **Prettier** rules:

  - [x] **printWidth** should be around **140** characters per line.

- [ ] Improve Code Coverage to reach 100%.

- [ ] Migrate [original documentation](https://github.com/robotlegs/robotlegs-framework/blob/master/src/readme.md) and adapt it to TypeScript and Pixi.

## [Unreleased]

<!--
Types of changes:

#### Added
- for new features.

#### Changed
- for changes in existing functionality.

#### Deprecated
- for soon-to-be removed features.

#### Removed
- for now removed features.

#### Fixed
- for any bug fixes.

#### Security
- in case of vulnerabilities.
-->

#### Changed

- Update dev dependencies to latest version.

#### Security

- Bump acorn to solve security vulnerability (see #133).

## Robotlegs-Pixi 2.0.0

### [v2.0.0](https://github.com/RobotlegsJS/RobotlegsJS-Pixi/releases/tag/2.0.0) - 2020-03-03

#### Breaking Change

- Update `pixi.js` to version `^5.0.0` (see #122).

  - `eventemitter3` package is not necessary anymore, you should remove this dependency from your `package.json`.

  - `@types/pixi.js` package is not necessary anymore, you should remove this dependency from your `package.json`.

  - For compatibility with version 4 of `pixi.js`, use `@robotlegsjs/pixi` version 1.

- Update [`@robotlegsjs/core`](https://github.com/RobotlegsJS/RobotlegsJS) to version `^2.0.0` (see #122 and #131).

  - Migrate array notation from `Array<SomeType>` to `SomeType[]`.

  - The rest of the `Public API` remains unchanged.

#### Added

- Add **Tidelift** as funding option (see #126).

- Add **Enterprise Support** information (see #127).

- Deploy example project (see #130).

#### Changed

- Update `tslib` to version `1.11.1` (see #129).

- Update dev dependencies to latest version.

#### Security

- Migrate to [terser-webpack-plugin](https://github.com/webpack-contrib/terser-webpack-plugin) to solve security vulnerability (see #125).

## Robotlegs-Pixi 1.0.0

### [v1.0.1](https://github.com/RobotlegsJS/RobotlegsJS-Pixi/releases/tag/1.0.1) - 2019-10-24

#### Changed

- Update `@robotlegsjs/core` to version `1.0.3` (see #117).

- Update `pixi.js` to version `4.8.8`.

- Remove `instanbul` dependency (see #116).

- Improve `prettier` rules and `autoformat` script (see #105).

- Enable `"editor.formatOnSave"` rule for `VS Code` (see #105).

- Migrate project to `travis-ci.com`.

- Update `codebeat` Project UUID.

- Update dev dependencies to latest version.

### [v1.0.0](https://github.com/RobotlegsJS/RobotlegsJS-Pixi/releases/tag/1.0.0) - 2018-11-25

#### Changed

- Update `@robotlegsjs/core` to version `1.0.0` (see #97).

- Update `karma` setup to generate code coverage report only for `src` folder (see #77).

- Improve `webpack` configuration used to run example project. The `npm start` script will generate hashed files (to avoid browser cache) and open the broswer automatically (see #78).

- Migrate to Headless Chrome and improve performance of `karma` (see #93).

- Prepare package for stable version (see #94).

- Update GitHub Templates (see #96).

- Update dev dependencies to latest version.

## Robotlegs-Pixi 0.2.0

### [v0.2.0](https://github.com/RobotlegsJS/RobotlegsJS-Pixi/releases/tag/0.2.0) - 2018-08-02

#### Changed

- Update `@robotlegsjs/core` to version `0.2.0` (see #74).

- Move `pixi.js` and `eventemitter3` libraries to **peerDependencies** (see #76).

- Enforce TSLint rules (see #58).

- Update TypeScript Compiler Options (see #57, #71).

- Use [tslib](https://github.com/Microsoft/tslib) library to avoid duplicated declarations (see #71).

- Update codeclimate config to version 2 (see #56).

- Use `rimraf` instead of `rm -rf` (see #55).

- Update Prettier rules (see #54).

- Adopts year-agnostic copyright message (see #70).

- Update dev dependencies to latest version.

#### Fixed

- Fix example (see #60).

- Only dispatch `added` or `removed` events when view is related to current stage (see #67, #68, #72).

## Robotlegs-Pixi 0.1.0

### [v0.1.2](https://github.com/RobotlegsJS/RobotlegsJS-Pixi/releases/tag/0.1.2) - 2017-11-23

#### Added

- Add `ContextView` handler (see #35).

#### Changed

- Update README (see #36).

- Update dev dependencies to latest version (see #36).

### [v0.1.1](https://github.com/RobotlegsJS/RobotlegsJS-Pixi/releases/tag/0.1.1) - 2017-11-20

#### Changed

- Update `pixi.js` to version `4.6.1` (see #33).

- Validate `IContextView` on extensions that depend on it (see #34).

- Update dev dependencies to latest version.

### [v0.1.0](https://github.com/RobotlegsJS/RobotlegsJS-Pixi/releases/tag/0.1.0) - 2017-11-13

#### Added

- Add changelog (see #27).

- Add code of conduct (see #28).

- Add issue template (see #29).

- Add pull request template (see #30).

#### Changed

- View property in **IContextView** is a **PIXI.Container** (see #26).

- `ContextView` is initialised with a reference to the stage (see #26).

- Helper methods on **Mediator** class are updated (see #26).

- Internal dependencies to PIXI are explicit (see #26).

- Map methods are used properly (see #26).

- RobotlegsJS/Core is migrated to version 0.1.1 (see #26 and #31).

- Improve Code Coverage and Fix Bugs (see #26).

- Update dev dependencies to latest version.

#### Removed

- `StageSyncExtension` class is removed (see #26).

#### Fixed

- Fix patch for PIXI and add contains method (see #26).


example/Game.ts
// ------------------------------------------------------------------------------
//  Copyright (c) 2017-present, RobotlegsJS. All Rights Reserved.
//
//  NOTICE: You are permitted to use, modify, and distribute this file
//  in accordance with the terms of the license agreement accompanying it.
// ------------------------------------------------------------------------------

import PIXI = require("pixi.js");

import { Context, MVCSBundle } from "@robotlegsjs/core";
import { ContextView, PixiBundle } from "../src";

import { MyConfig } from "./config/MyConfig";
import { RobotlegsView } from "./view/RobotlegsView";

export class Game {
    private canvas: HTMLCanvasElement;
    private stage: PIXI.Container;
    private renderer: PIXI.Renderer;
    private context: Context;

    constructor() {
        this.canvas = <HTMLCanvasElement>document.getElementById("canvas");
        this.renderer = PIXI.autoDetectRenderer({
            width: 960,
            height: 400,
            view: this.canvas,
            backgroundColor: 0xffffff
        });
        this.stage = new PIXI.Container();

        this.context = new Context();
        this.context
            .install(MVCSBundle, PixiBundle)
            .configure(new ContextView(this.stage))
            .configure(MyConfig)
            .initialize();

        this.stage.addChild(new RobotlegsView());

        document.body.appendChild(this.renderer.view);

        this.render();
    }

    public render = () => {
        this.renderer.render(this.stage);
        window.requestAnimationFrame(this.render);
    };
}
